### PR TITLE
[ConstraintSystem] Detect and diagnose missing generic arguments

### PR DESCRIPTION
**Pull request by xedin**
Introduce a fix to detect and diagnose situations when omitted
generic arguments couldn't be deduced by the solver based on
the enclosing context.

Example:

```swift
struct S<T> {
}

_ = S() // There is not enough context to deduce `T`
```

Resolves: rdar://problem/51203824

### CODE REVIEW

File: test/Constraints/array_literal.swift [hunk 1]
```diff
@@ -102,7 +102,7 @@ func longArray() {
   var _=["1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1"]
 }
 
-[1,2].map // expected-error {{expression type '((Int) throws -> _) throws -> [_]' is ambiguous without more context}}
+[1,2].map // expected-error {{generic parameter 'T' could not be inferred}}
```
Comment (slavapestov) [inline test/Constraints/array_literal.swift]: Nice!

File: test/Generics/invalid.swift [hunk 2]
```diff
@@ -79,7 +79,8 @@ func badDiagnostic2() {
 func takesAny(_ a: Any) {}
 
 func badDiagnostic3() {
-  takesAny(Deli.self) // expected-error {{argument type 'Deli<_>.Type' does not conform to expected type 'Any'}}
```
Comment (slavapestov) [inline test/Generics/invalid.swift]: Nice!

File: test/decl/nested/type_in_type.swift [hunk 3]
```diff
@@ -398,11 +398,14 @@ func test() {
...
+  // expected-error@-1 {{generic parameter 'A' could not be inferred}}
+  // expected-note@-2 {{explicitly specify the generic arguments to fix this issue}} {{36-36=<<#A: ExpressibleByCatLiteral#>>}}
   let _: Claws.Fangs<NotADog> = something()
-  // expected-error@-1 {{generic parameter 'T' could not be inferred}} // FIXME: bad diagnostic
-  _ = Claws.Fangs<NotADog>() // TODO(diagnostics): There should be two errors here - "cannot infer A" and "NotADog conformance to ExpressibleByDogLiteral"
-  // expected-error@-1 {{referencing initializer 'init()' on 'Claws.Fangs' requires that 'NotADog' conform to 'ExpressibleByDogLiteral'}}
+  // expected-error@-1 {{generic parameter 'A' could not be inferred}}
+  _ = Claws.Fangs<NotADog>()
+  // expected-error@-1 {{generic parameter 'A' could not be inferred}}
+  // expected-error@-2 {{referencing initializer 'init()' on 'Claws.Fangs' requires that 'NotADog' conform to 'ExpressibleByDogLiteral'}}
```
Comment (slavapestov) [inline test/decl/nested/type_in_type.swift]: Nice!

File: test/decl/typealias/generic.swift [hunk 4]
```diff
@@ -72,17 +72,18 @@ typealias G<S1, S2> = A<S1, S2>
...
+let _ : D = D(a: 1, b: 2)
+// expected-error@-1 {{generic parameter 'T3' could not be inferred}}
+// expected-note@-2 {{explicitly specify the generic arguments to fix this issue}} {{14-14=<Int, Int, Any>}}
 
 let _ : D<Int, Int, Float> = D<Int, Int, Float>(a: 1, b: 2)
 
-// FIXME: This is not a great error.
-// expected-error @+1 {{cannot convert value of type 'D<Int, Int, Float>' (aka 'MyType<Int, Int>') to specified type 'D'}}
 let _ : D = D<Int, Int, Float>(a: 1, b: 2)
+// expected-error@-1 {{generic parameter 'T3' could not be inferred}}
```
Comment (slavapestov) [inline test/decl/typealias/generic.swift]: Nice!

**Review by slavapestov — APPROVED** [review]
Great stuff!

File: lib/Sema/CSDiagnostics.cpp [hunk 5]
```diff
@@ -2881,3 +2885,235 @@ MissingContextualConformanceFailure::getDiagnosticFor(
...
+
+  if (!diagnosed)
+    return false;
+
+  auto *DC = getDeclContext();
+  if (auto *AFD = dyn_cast<AbstractFunctionDecl>(DC)) {
+    if (isa<ConstructorDecl>(AFD)) {
+      emitDiagnostic(AFD, diag::note_call_to_initializer);
+    } else {
+      emitDiagnostic(AFD,
```
Comment (slavapestov) [inline lib/Sema/CSDiagnostics.cpp]: What about generic subscripts?

File: lib/Sema/CSDiagnostics.cpp [hunk 6]
```diff
@@ -2881,3 +2885,235 @@ MissingContextualConformanceFailure::getDiagnosticFor(
...
+  Type baseTyForNote;
+  auto *DC = getDeclContext();
+
+  if (auto *NTD =
+          dyn_cast_or_null<NominalTypeDecl>(DC->getSelfNominalTypeDecl())) {
+    baseTyForNote = NTD->getDeclaredType();
+  } else if (auto *TAD = dyn_cast<TypeAliasDecl>(DC)) {
+    baseTyForNote = TAD->getUnboundGenericType();
+  } else {
+    baseTyForNote = DC->getDeclaredInterfaceType();
```
Comment (slavapestov) [inline lib/Sema/CSDiagnostics.cpp]: This is unreachable

**Review by slavapestov — COMMENTED** [review]
Sorry I do have a couple of minor comments :)

File: test/Constraints/generics.swift [hunk 7]
```diff
@@ -382,32 +384,47 @@ func testFixItNested() {
   _ = [FullyGeneric]() // expected-error {{generic parameter 'Foo' could not be inferred}} expected-note {{explicitly specify the generic arguments to fix this issue}} {{20-20=<Any>}}
 
   _ = FullyGeneric<FullyGeneric>() // expected-error {{generic parameter 'Foo' could not be inferred}}
+  // expected-note@-1 {{explicitly specify the generic arguments to fix this issue}} {{32-32=<Any>}}
 
-  _ = Pair< // expected-error {{generic parameter 'Foo' could not be inferred}}
+  _ = Pair<
     FullyGeneric,
```
Comment (xedin) [inline test/Constraints/generics.swift]: @slavapestov I think the biggest improvement here is actually related to couple of examples down here, which we now diagnostic relative to TypeRepr and suggest all of the missing signatures.

**Comment by jckarter** [comment]
This is great! Would it be possible to show the declaration name in the diagnostic? For some of these cases, I think it'd be easier to understand if the diagnostic could point at the declaration the generic argument comes from, e.g.:

```
error: generic parameter 'T' of 'map' could not be inferred
note: 'map' declared here
```

**Comment by xedin** [comment]
@jckarter Yes, absolutely, it's just a better of getting more cases for the note, it currently only covers functions and initializes, I'll look into that.

File: lib/Sema/CSDiagnostics.cpp (hunk 6 above)
Comment (xedin) [reply lib/Sema/CSDiagnostics.cpp]: Fixed!

File: lib/Sema/CSDiagnostics.cpp (hunk 5 above)
Comment (xedin) [reply lib/Sema/CSDiagnostics.cpp]: Added a note for generic subscripts, thanks!

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Pull request merged**